### PR TITLE
drivers: sensor: add mutex to cmd_get_sensor()

### PR DESCRIPTION
**Pull request by marco85mars20**
Add a mutex to protect shared data-structures, since shell can have multiple backends.

### CODE REVIEW

File: drivers/sensor/sensor_shell.c [hunk 1]
```diff
@@ -299,8 +303,15 @@ static void sensor_shell_processing_callback(int result, uint8_t *buf, uint32_t
 static int cmd_get_sensor(const struct shell *sh, size_t argc, char *argv[])
 {
 	const struct device *dev;
+	int count = 0;
 	int err;
 
+	err = k_mutex_lock(cmd_get_mutex, K_NO_WAIT);
```
Comment (yperess) [inline drivers/sensor/sensor_shell.c:309]: There are a few return statements below that will need to unlock this mutex

**Review by yperess — CHANGES_REQUESTED** [review]
Can you describe the issue in more detail. I see that you wrote that the shell might have more than one backend. Can we run more than one command at a time? If I'm reading this correctly, the concern is that the iodev_sensor_shell_read_config will change during the async read? If so, then the unlock should go in the error cases and for the success, in the callback

File: drivers/sensor/sensor_shell.c (hunk 1 above)
Comment (marco85mars20) [reply drivers/sensor/sensor_shell.c]: I agree, my bad. I had moved the lock to the top of the function, but originally it was placed at the very bottom. And now I have moved it back again.

**Comment by marco85mars20** [comment]
@yperess The shell can have multiple backends enabled at the same time, in my case, I use for instance Bluetooth, mcumgr, and either serial or RTT. This could potentially cause misbehaviors in case `sensor get` is executed on multiple backends at overlapping times. The root cause of the issue is using a set of global variables (ie. `iodev_sensor_shell_channels`, `iodev_sensor_shell_read_config`, `sensor_read_rtio`) in `cmd_get_sensor()`. And at the moment there is no way to guarantee that the content of these variables is not changing during a get operation.

**Comment by yperess** [comment]
> @yperess The shell can have multiple backends (AKA "transports") enabled at the same time, in my case, I use for instance Bluetooth, mcumgr, and either serial or RTT. Each backend runs on its own thread created by `shell_init()` and each thread can execute 1 command at a time, thus the same command could be executed on different transports at the same time.

Thanks for the details. Let me look into the backend API today. It would be better if we had an rtio device per backend. Is there a macro to do that? Or kconfig for the count of backends?

**Review by yperess — DISMISSED** [review]
This works, I'll deal with optimizing later.

**Review by nixward — CHANGES_REQUESTED** [review]
@marco85mars20 the mutex lock needs to happen before `iodev_sensor_shell_channels[]` is updated because that array is used in the `sensor_read()` function.

**Comment by nixward** [comment]
@marco85mars20 thinking about this a bit more, this will likely be a problem not just for this shell command. Should shell backends lock the shell while using it? It would mean any long running shell commands would block other backends from running shell commands but it would require only one mutex.

**Review by yperess — APPROVED** [review]

**Comment by marco85mars20** [comment]
@nixward sorry for the late reply.
This specific shell command reuses the same data structure for potentially concurrent operations. I haven't seen this issue in any other shell commands.

**Review by teburd — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**